File: notes_app/lib/services/note_service.dart
import 'package:appwrite/appwrite.dart';
import 'package:appwrite/models.dart';
import '../config/appwrite_config.dart';

class NoteService {
  late final Databases _databases;

  NoteService() {
    _databases = Databases(AppwriteConfig.client);
  }

  Future<List<Document>> getNotes({String? userId}) async {
    try {
      List<String> queries = [];
      if (userId != null) queries.add(Query.equal('user_id', userId));
      queries.add(Query.orderDesc('createdAt'));

      final response = await _databases.listDocuments(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        queries: queries,
      );
      return response.documents;
    } catch (e) {
      rethrow;
    }
  }

  Future<Document> createNote(Map<String, dynamic> data) async {
    try {
      final noteData = {
        ...data,
        'createdAt': DateTime.now().toIso8601String(),
        'updatedAt': DateTime.now().toIso8601String(),
      };
      return await _databases.createDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: ID.unique(),
        data: noteData,
      );
    } catch (e) {
      rethrow;
    }
  }

  Future<bool> deleteNote(String noteId) async {
    try {
      await _databases.deleteDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: noteId,
      );
      return true;
    } catch (e) {
      rethrow;
    }
  }

  Future<Document> updateNote(String noteId, Map<String, dynamic> data) async {
    try {
      final noteData = { ...data, 'updatedAt': DateTime.now().toIso8601String() };
      return await _databases.updateDocument(
        databaseId: AppwriteConfig.databaseId,
        collectionId: AppwriteConfig.collectionId,
        documentId: noteId,
        data: noteData,
      );
    } catch (e) {
      rethrow;
    }
  }
}

File: NotesApp/src/services/note.service.js
import { Databases, ID, Query } from "appwrite";
import client from "../config/appwrite";
import { APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID } from "@env";

const databases = new Databases(client);

export const getNotes = async (userId = null) => {
  try {
    const queries = [];
    if (userId) {
      queries.push(Query.equal("user_id", userId));
    }
    queries.push(Query.orderDesc("createdAt"));
    const response = await databases.listDocuments(APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID, queries);
    return response.documents;
  } catch (error) {
    console.error("Error getting notes:", error);
    throw error;
  }
};

export const createNote = async (data) => {
  try {
    const noteData = {
      ...data,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
    const response = await databases.createDocument(
      APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID, ID.unique(), noteData
    );
    return response;
  } catch (error) {
    console.error("Error creating note:", error);
    throw error;
  }
};

export const deleteNote = async (noteId) => {
  try {
    await databases.deleteDocument(APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID, noteId);
    return true;
  } catch (error) {
    console.error("Error deleting note:", error);
    throw error;
  }
};

export const updateNote = async (noteId, data) => {
  try {
    const noteData = { ...data, updatedAt: new Date().toISOString() };
    const response = await databases.updateDocument(
      APPWRITE_DATABASE_ID, APPWRITE_COLLECTION_ID, noteId, noteData
    );
    return response;
  } catch (error) {
    console.error("Error updating note:", error);
    throw error;
  }
};
